### DISCUSSION TITLE
如何优先挂载messageApi

### DISCUSSION THREAD

**Discussion by seafronthu** [umijs/umi#10504] [Q&A] [answered]
期望：我的要求是用户信息需要提前调用接口拿到手，它是用来判断是否登录的，但是接口如果报错我又想通过挂载在全局变量的messageApi来提醒用户。
问题：现在getInitialState中不能挂载messageApi。
我该如何处理

**Comment by fz6m** [comment, marked as answer]
全局 layout 里取一下 initial state 或者用全局状态管理库（比如 [valtio](https://umijs.org/docs/max/valtio) ），发现未登录的话就弹 message 。

**Reply by seafronthu** [reply]
可能我没有描述清楚。有些接口必须在getInitialState先调用完成，在getInitialState中调用接口的过程中有部分接口报错了，我得提示错误信息给用户，但是呢getInitialState不能使用messageApi挂载到全局的对象中，也就是我不能在getInitialState使用message.useMessage

**Reply by fz6m** [reply]
在 `rootContainer` 上一层 provider ，再把 messageApi 存在 valtio 里用吧，valtio 可以无视 hooks 。

**Reply by seafronthu** [reply]
是的，在运行时配置加个rootContainer函数，里面封装个Provider组件挂载useMessage就行了，完美。我还一直以为getInitialState这个优先级最高。万分感谢！！！

**Reply by seafronthu** [reply]
这是我的方案：
```javascript
import { Provider } from "@umijs/max"
import React from 'react'
import { message, notification } from 'antd';
import globalFn from '@/utils/global';
const ProviderContainer: React.FC<{ children?: React.ReactNode }> = props => {
  const [messageApi] = message.useMessage();
  const [notificationApi] = notification.useNotification();
  globalFn.bindTarget('messageApi', messageApi);
  globalFn.bindTarget('notificationApi', notificationApi);
  const { children } = props
  const modes = { "a": () => { } }
  return <Provider models={modes}>{children}</Provider>
}
export default ProviderContainer
```
